umbral/keys.py

```diff
@@ -1,472 +1,205 @@
-from cryptography.exceptions import InternalError
-from cryptography.hazmat.primitives import hashes
-from cryptography.hazmat.primitives.kdf.hkdf import HKDF
-from cryptography.hazmat.primitives.kdf.scrypt import Scrypt as CryptographyScrypt
-from nacl.secret import SecretBox
+from typing import TYPE_CHECKING, Tuple
 
-from umbral import openssl
-from umbral.config import default_params
-from umbral.curvebn import CurveBN
-from umbral.params import UmbralParameters
-from umbral.point import Point
-from umbral.curve import Curve
-from umbral.random_oracles import hash_to_curvebn
+from cryptography.exceptions import InvalidSignature
+from cryptography.hazmat.primitives.asymmetric import utils
+from cryptography.hazmat.primitives.asymmetric.ec import ECDSA
 
+from . import openssl
+from .curve import CURVE
+from .curve_scalar import CurveScalar
+from .curve_point import CurvePoint
+from .dem import kdf
+from .serializable import Serializable
 
-__SALT_SIZE = 32
+if TYPE_CHECKING: # pragma: no cover
+    from .hashing import Hash
 
 
-class Scrypt:
-    __DEFAULT_SCRYPT_COST = 20
-
-    def __call__(self,
-                 password: bytes,
-                 salt: bytes,
-                 **kwargs) -> bytes:
-        """
-        Derives a symmetric encryption key from a pair of password and salt.
-        It also accepts an additional _scrypt_cost argument.
-        WARNING: RFC7914 recommends that you use a 2^20 cost value for sensitive
-        files. It is NOT recommended to change the `_scrypt_cost` value unless
-        you know what you are doing.
-        :param password: byte-encoded password used to derive a symmetric key
-        :param salt: cryptographic salt added during key derivation
-        :return:
-        """
-
-        _scrypt_cost = kwargs.get('_scrypt_cost', Scrypt.__DEFAULT_SCRYPT_COST)
-        try:
-            derived_key = CryptographyScrypt(
-                salt=salt,
-                length=SecretBox.KEY_SIZE,
-                n=2 ** _scrypt_cost,
-                r=8,
-                p=1,
-                backend=default_backend()
-            ).derive(password)
-        except InternalError as e:
-            required_memory = 128 * 2**_scrypt_cost * 8 // (10**6)
-            if e.err_code[0].reason == 65:
-                raise MemoryError(
-                    "Scrypt key derivation requires at least {} MB of memory. "
-                    "Please free up some memory and try again.".format(required_memory)
-                )
-            else:
-                raise e
-        else:
-            return derived_key
-
-
-def derive_key_from_password(password: bytes,
-                             salt: bytes,
-                             **kwargs) -> bytes:
+class SecretKey(Serializable):
     """
-    Derives a symmetric encryption key from a pair of password and salt.
-    It uses Scrypt by default.
+    Umbral secret (private) key.
```

Comment:
Usually the term "private" key implicitly signals an asymmetric cryptography scheme (hinting at the private vs public dichotomy), while "secret" key is usually associated to symmetric cryptography.

Reply:
I was following the name convention from `rust-umbral`, which, in turn, follows the terms used by the RustCrypto stack. Is using "secret" instead of "private" here very confusing, or just somewhat unusual?

As an extension of uniformity with RustCrypto, I was thinking of just re-exporting the RustCrypto key objects themselves instead of wrapping them in our own types. This way the keys can be used with other libraries based on RustCrypto, making it unnecessary to have conversion methods like `to_cryptography_privkey()` we have in PyUmbral.

Also being able to abbreviate to "sk" and "pk" is quite useful.